codex/stores/maintenance.nim

```diff
@@ -10,90 +10,318 @@
+        md.manifestsCids = (currDatasetMd.manifestsCids & manifestsCids).deduplicate
+        md.checkpoint = Checkpoint(progress: 0, timestamp: 0)
+      else:
+        md.expiry = expiry
+        md.leavesCount = leavesCount
+        md.manifestsCids = manifestsCids
+        md.checkpoint = Checkpoint(progress: 0, timestamp: 0)
+
+      md.some
+  )
+
+proc trackExpiry*(
+  self: DatasetMaintainer,
+  treeCid: Cid,
+  leavesCount: Natural,
+  manifestsCids: seq[Cid] = @[]
+): Future[?!void] {.async.} =
+  await self.trackExpiry(treeCid, leavesCount, (await self.clock.now) + self.defaultExpiry.seconds, manifestsCids)
+
+proc trackExpiry*(
+  self: DatasetMaintainer,
+  cid: Cid,
+  manifestsCids: seq[Cid] = @[]
+): Future[?!void] {.async.} =
+  await self.trackExpiry(cid, 0, (await self.clock.now) + self.defaultExpiry.seconds, manifestsCids)
+
+proc ensureExpiry*(
+  self: DatasetMaintainer,
+  treeCid: Cid,
+  minExpiry: SecondsSince1970): Future[?!void] {.async.} =
+  ## Sets the dataset expiry to a maximum of current expiry and `minExpiry`,
+  ## if a dataset for given `treeCid` is not currently tracked a CatchableError is thrown
+  ##
+
+  trace "Updating a dataset expiry", treeCid, minExpiry
+
+  without key =? createDatasetMetadataKey(treeCid), err:
+    return failure(err)
+
+  await self.metaDs.modifyT(key,
+    proc (maybeCurrDatasetMd: ?DatasetMetadata): Future[?DatasetMetadata] {.async.} =
+      if currDatasetMd =? maybeCurrDatasetMd:
+        let datasetMd = DatasetMetadata(
+          expiry: max(currDatasetMd.expiry, minExpiry),
+          leavesCount: currDatasetMd.leavesCount,
+          manifestsCids: currDatasetMd.manifestsCids,
+          checkpoint: currDatasetMd.checkpoint
+        )
+        return datasetMd.some
+      else:
+        raise newException(CatchableError, "DatasetMetadata for treeCid " & $treeCid & " not found")
+  )
+
+
+proc recordCheckpoint*(self: DatasetMaintainer, treeCid: Cid, datasetMd: DatasetMetadata): Future[?!void] {.async.} =
+  # Saves progress or deletes dataset metadata if progress > leavesCount
+  #
 
-proc runBlockCheck(self: BlockMaintainer): Future[void] {.async.} =
-  let expirations = await self.repoStore.getBlockExpirations(
-    maxNumber = self.numberOfBlocksPerInterval,
-    offset = self.offset
+  without key =? createDatasetMetadataKey(treeCid), err:
+    return failure(err)
+
+  await self.metaDs.modifyT(key,
+    proc (maybeCurrDatasetMd: ?DatasetMetadata): Future[?DatasetMetadata] {.async.} =
+      if currDatasetMd =? maybeCurrDatasetMd:
+        if currDatasetMd.expiry != datasetMd.expiry or currDatasetMd.manifestsCids != datasetMd.manifestsCids:
+          raise newException(CatchableError, "Change in expiry detected, interrupting maintenance for dataset with treeCid " & $treeCid)
+
+        if currDatasetMd.checkpoint.progress > datasetMd.checkpoint.progress:
+          raise newException(CatchableError, "Progress should be increasing only, treeCid " & $treeCid)
```

Comment:
`modifyT` is a wrapper for `modify` which is a function that's a shorthand for `modifyGet` ([link to docs](https://github.com/codex-storage/nim-datastore/blob/d02f2e7380410137fd2cdc9bd0d06c0004f7e8ff/datastore/datastore.nim#L50)). Closure in `modifyT` is of type `?T -> Future[?T]`. If we would add a result to it, it would complicate the signature significantly, it would be then `?T -> Future[Result[?T]]`, while `modifyTGetU` would be `?T -> Future[Result[(?T, U)]]`. Instead of this complication closure is expected to throw `CatchableError` if error happens. I think it makes sense to keep it this way. Wdyt?

Reply:
I think @dryajov covered this in his comment regarding `modifyGet` 👍